### ISSUE TITLE
Database actions have extra transaction which cause a lock (Magento EE 2.3.3)

### ISSUE THREAD

**Issue by rikwillems** [netz98/n98-magerun2#502]
I've created a console command that updates the status of an order. When executing the command an extra transaction is started that causes a lock. The extra transaction is not there when using `bin/magento`.

The transaction is started by this line: `vendor/magento/module-customer-custom-attributes/Model/Sales/AbstractSales.php`

Steps to reproduce:
* use Magento Commerce 2.3.3
* place an order
* Install: https://github.com/rikwillems/Magento-2-sample-order-unhold-console-command
* execute command
    * `bin/magento sample:order:hold <order_id>`
    * `bin/magento sample:order:unhold <order_id>`
    * `n98-magerun2.phar sample:order:hold <order_id>`
    * `n98-magerun2.phar sample:order:unhold <order_id>`

The n98-magerun2 commands will end with an exception. In your MySQL process monitor you can see the lock being requested.

**Comment by cmuench** [comment]
@rikwillems We encountered that problem in some 3rd party Magento modules as well. In Magento 2.3.3 the database layer was changed under the hood.
One of my colleagues had the same issue with a product import started by a CLI command.
If I remember correct, the `$appState->setAreaCode` is opening the transaction.
I am currently not sure how we can solve that issue. If we remove that code other commands will fail. Thanks for reporting. This needs more investigation.

**Comment by rikwillems** [comment]
@cmuench please let me know if I can be of any assistance.

Without `$appState->setAreaCode` the problem also occurs for me.

**Comment by ktomk** [comment]
@rikwillems Can you add the exception to the report?

**Comment by rikwillems** [comment]
@ktomk The process monitor shows query: `SELECT `sales_order`.`entity_id` FROM `sales_order` WHERE (entity_id = '539597') FOR UPDATE`

The command line output shows:
```
[Magento\Framework\DB\Adapter\LockWaitException]
  SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: SELECT `sales_order`.`entity_id` FROM `sales_order`
   WHERE (entity_id = '539597') FOR UPDATE

  [Zend_Db_Statement_Exception]
  SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: SELECT `sales_order`.`entity_id` FROM `sales_order`
   WHERE (entity_id = '539597') FOR UPDATE

  [PDOException]
  SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction
```

The exception.log file show: 
```
2019-12-20 12:20:21] main.CRITICAL: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: SELECT `sales_order`.`entity_id` FROM `sales_order` WHERE (entity_id = '539597') FOR UPDATE {"exception":"[object] (Magento\\Framework\\DB\\Adapter\\LockWaitException(code: 1205): SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: SELECT `sales_order`.`entity_id` FROM `sales_order` WHERE (entity_id = '539597') FOR UPDATE at vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:593, Zend_Db_Statement_Exception(code: 0): SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: SELECT `sales_order`.`entity_id` FROM `sales_order` WHERE (entity_id = '539597') FOR UPDATE at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:110, PDOException(code: HY000): SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction at vendor/magento/framework/DB/Statement/Pdo/Mysql.php:91)"} []
```

**Comment by ktomk** [comment]
So if the db-layer was still the old one of ZF1 previously IIRC there are "stacked" transactions, so technically it was possible to start a database transaction even one was already running. The db layer just counted up on transaction start and counted down on transaction end. Only when the counter was 0 the transaction was actually started (or stopped).

Maybe this feature is now gone with the new db layer under the hood and now it creates a race condition.

I just guess this out of thin air, I have not looked into the new db code @cmuench was mentioning.

Apart from the db-layer this might highlight the problem in the application run lifecycle we sometimes have w/ magerun as the magento application is normally initialized pretty early when a magerung command is run (IIRC). At least it is sometimes hard to find the right spot where/when to do bootstraping / initialization. E.g. when needing the database layer, at least some part of the Magento application needs to be created to aquire it.

@rikwillems Is it that you could use `bin/magento` here and it does the job for you and it's just that `n98-magerun2.phar` chokes on these commands?

**Comment by rikwillems** [comment]
Yes @ktomk it works fine with `bin/magento` and not with `n98-magerun2.phar`.
So we can work around it. But it cost me quite some time to think of the relation with n98-magerun2. Let's hope this issue saves others the same investigation.

**Comment by ktomk** [comment]
Thanks for your feedback, this brings me the idea that these commands could be shell'ed out in a new php subprocess. When we get this working it would be a good test to see if magerun2 can initialize and run commands w/o affecting the runtime. E.g. if it's the database layer here. Not knowing for sure Symfony Console brings the correct abstractions to do this, but it's maybe worth a try.

**Comment by cmuench** [comment]
@rikwillems I installed your module in a vanilla Magento Open Source 2.3.3 and everything works fine with n98-magerun2 v2.3.3.
On my machine the behaviour is currently not reproducible, yet. In the next version I will remove the "var" directory check during bootstrap. This check is currently forcing an very early Magento initialization which I want to prevent. Maybe this is also a problem in your case.

**Comment by rikwillems** [comment]
@cmuench this error occurs in Enterprise Edition not in Open Source.

**Comment by cmuench** [comment]
@rikwillems ok. Sorry. I will try out that in a fresh Commerce Edition, too.

**Comment by cmuench** [comment]
@rikwillems The good thing... I can reproduce it now.
The problem is a event which is loaded.
The problematic line is this: 
vendor/magento/module-customer-custom-attributes/Model/Sales/AbstractSales.php +72

I have currently no clue what's wrong.

**Comment by rikwillems** [comment]
@cmuench Yes that is the case. It opens a new transaction for some reason. I've had an issue with n98 before on a command that did work with `bin/magento`. Might have been the same kind of thing.

I only run n98 through the phar. How can I best run the normal php code to debug such a thing?

**Comment by cmuench** [comment]
@rikwillems I use a cloned n98-magerun2 version with an alias to to the shipped bin.

e.g.

`alias n98-magerun2-dev='php ~/Workspaces/opensource/n98-magerun2/bin/n98-magerun2'`

This works also with a debugger. I never tried to debug the phar file.

**Comment by cmuench** [comment]
@rikwillems I did some changes in the bootstrap process. In my test system (EE 2.3.3) the error is gone.

Can you please verify that on your machine?

`n98-magerun2.phar self-update --unstable`

**Comment by rikwillems** [comment]
@cmuench it's running `4.0.0` now but I still have the same issue.

**Comment by cmuench** [comment]
@rikwillems I have started my MySQL with `--transaction-isolation=READ-COMMITTED`.
We have set transaction isolation on all servers to READ-COMMITTED. Since this most of the transaction problems are gone on production servers. Maybe that's the pupose that I did not face the problem.

**Comment by cmuench** [comment]
Possible relation to this: #667

**Comment by cmuench** [comment]
@rikwillems I found an issue that we bootstrapped twice. That's fixed. Maybe it's related to this issue.
I just installed your test module in a Magento 2.3.7-p3 and got no error with the latest n98-magerun2 v5.1.0.

**Comment by rikwillems** [comment]
That is great news, good find! Thanks for keeping an eye of this this must be an edge case.

**Comment by rikwillems** [comment]
I'm afraid I have to report, on another case, the issue is still present in 5.1.0 and 5.1.0-dev. When using `bin/magento` on the same console command there is no issue. Please find my command below, it copies product attribute values from one store view to another.

```
<?php

declare(strict_types=1);

namespace Company\Configuration\Console\Command;

use Magento\Catalog\Api\ProductRepositoryInterface;
use Magento\Catalog\Model\ProductFactory;
use Magento\Framework\Api\SearchCriteriaBuilder;
use Magento\Framework\App\Area;
use Magento\Framework\App\ResourceConnection;
use Magento\Framework\App\State;
use Magento\Store\Api\Data\StoreInterface;
use Magento\Store\Model\StoreManagerInterface;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Helper\ProgressBar;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

/**
 *
 */
class CopyProducts extends Command
{
    const IMPORT_BATCH_SIZE = 200;

    private ResourceConnection $resourceConnection;
    private ProductRepositoryInterface $productRepository;
    private ProductFactory $productFactory;
    private SearchCriteriaBuilder $searchCriteriaBuilder;
    private State $appState;
    private StoreManagerInterface $storeManager;

    public function __construct(
        ResourceConnection $resourceConnection,
        ProductRepositoryInterface $productRepository,
        SearchCriteriaBuilder $searchCriteriaBuilder,
        ProductFactory $productFactory,
        State $appState,
        StoreManagerInterface $storeManager
    ) {
        parent::__construct();
        $this->resourceConnection = $resourceConnection;
        $this->productRepository = $productRepository;
        $this->productFactory = $productFactory;
        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
        $this->appState = $appState;
        $this->storeManager = $storeManager;
    }

    /** {@inheritdoc} */
    protected function configure()
    {
        $this
            ->setName('company:copy:products')
            ->setDescription('Copy products to store - one time use only')
            ->addArgument('from_store', InputArgument::REQUIRED)
            ->addArgument('to_store', InputArgument::REQUIRED)
        ;
    }

    /** {@inheritdoc} */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $output->setDecorated(true);
        $this->appState->setAreaCode(Area::AREA_GLOBAL);

        $fromStore = $this->storeManager->getStore($input->getArgument('from_store'));
        $toStore = $this->storeManager->getStore($input->getArgument('to_store'));

        $connection = $this->resourceConnection->getConnection();
        $select = $connection->select()->from(
            $connection->getTableName('catalog_product_entity'),
            [new \Zend_Db_Expr('COUNT(entity_id)')]
        );

        $count = $connection->fetchOne($select);
        $batches = ceil($count / self::IMPORT_BATCH_SIZE);

        for ($batch = 1; $batch <= $batches; $batch++) {
            $output->writeln("\nBatch {$batch}/{$batches}");
            $this->importProducts($output, $batch, $fromStore, $toStore);
        }
    }

    private function importProducts(
        OutputInterface $output,
        int $batch,
        StoreInterface $fromStore,
        StoreInterface $toStore
    ) {
        $newWebsiteId = $toStore->getWebsiteId();
        $products = $this->productRepository->getList(
            $this->searchCriteriaBuilder->create()
                ->setPageSize(self::IMPORT_BATCH_SIZE)
                ->setCurrentPage($batch)
        );

        $progress = new ProgressBar($output, count($products->getItems()));
        $progress->setFormat('<comment>%message%</comment> %current%/%max% [%bar%] %percent:3s%% %elapsed%');

        foreach ($products->getItems() as $product) {
            $progress->setMessage("ID {$product->getId()} ");
            $websites = array_merge($product->getWebsiteIds(), [$newWebsiteId]); /** @phpstan-ignore-line */

            // @see \Magento\Catalog\Controller\Adminhtml\Product\Save->copyToStore()
            try {
                $this->productFactory->create()
                    ->setStoreId($fromStore->getId())
                    ->load($product->getId())
                    ->setStoreId($toStore->getId())
                    ->setCanSaveCustomOptions(true)
                    ->setCopyFromView(true)
                    ->setWebsiteIds($websites)
                    ->save()
                ;
            } catch (\Exception $e) {
                $output->writeln("<error>ID: {$product->getId()} - {$e->getMessage()}</error>");
            }

            $progress->advance();
        }
    }
}
```

**Comment by cmuench** [comment]
@rikwillems I rewrote the way the core commands are execute.
We will now call bin/magento in the background. That means that should not bootstrap Magento twice. I was able to remove the initialize call which was used to register the core commands in n98-magerun2.

There is a PR available with a phar file.
See: https://github.com/netz98/n98-magerun2/pull/1085#issuecomment-1385125684

The chance is very good that this fixes also the issue of the DB locks.

**Comment by cmuench** [comment]
The ticket is related to Magento 2.3.5 which has reached EOL.
For newer versions we hope the issue is fixed in the next release 7.0.0.
It can be tested by updating to current development version via `self-update --unstable`.
I close the issue. If it's still reproducible, please re-open.
